Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp

```diff
@@ -181,28 +181,40 @@ class PageLoaderClient final : public API::InjectedBundle::PageLoaderClient {
     {
         if (!frame.isMainFrame())
             return;
-        webkitWebPageSetURI(m_webPage, getDocumentLoaderURL(frame.coreFrame()->loader().provisionalDocumentLoader()));
+        const auto uri = getDocumentLoaderURL(frame.coreFrame()->loader().provisionalDocumentLoader());
+        auto webKitFrame = webkitFrameGetOrCreate(&frame);
+        webkitFrameSetUri(webKitFrame, uri);
```

Comment:
This only works for the main frame, we should update the uri for any frame. I don't think we want to create the WebKitFrame object just to set the uri, if it already exists we update the uri. What's the value of the frame url at this point? I think we could call webkitFrameUpdateURI() and there we always update the uri with the url value of the WebFrame object.

Reply:
You are right, I was concerned about using here webkitFrameGetOrCreate and create WebKitWebFrame object only for setting the uri as you mentioned. At this point the WebKitWebFrame does not exist and has to be created.

Reply:
So the idea here is to first check if WebKitWebFrame exists and only then update its uri, right?

Reply:
Right

Reply:
> What's the value of the frame url at this point?

unfortunately it is empty

> I think we could call webkitFrameUpdateURI() and there we always update the uri with the url value of the WebFrame object

what do you mean by webkitFrameUpdateURI? new function?

Reply:
It's empty because frame url always gets the document loader, not the provisional one. In that case we indeed need a SetURI.